tests/gp/test_hsgp_approx.py

```diff
@@ -165,11 +176,23 @@ def test_parametrization_drop_first(self, model, cov_func, X1, drop_first):
             else:
                 assert n_coeffs == n_basis, "one was dropped when it shouldn't have been"
 
-    @pytest.mark.parametrize("parameterization", ["centered", "noncentered"])
+    @pytest.mark.parametrize(
+        "cov_func,parameterization",
+        [
+            (pm.gp.cov.ExpQuad(1, ls=1), "centered"),
+            (pm.gp.cov.ExpQuad(1, ls=1), "noncentered"),
+            # (pm.gp.cov.Periodic(1, period=1, ls=1), None),
```

Comment:
I get passing tests when I uncomment this line, then zero the mean on `samples1` and `samples2`

```python
        samples1 = az.extract(idata.prior["f1"])["f1"].values.T
        samples2 = az.extract(idata.prior["f2"])["f2"].values.T

        samples1 = samples1 - np.mean(samples1, axis=1)[:, None]
        samples2 = samples2 - np.mean(samples2, axis=1)[:, None]
```

Does that work on your end?

Reply:
Yeah that passes for me. Is the logic sound for doing this? As in, is there a mathematical reason that we shouldn't do this?

Just double checking but it's all completely unrelated that we take the mean of the domain with `self.prior_linearized(X - self._X_mean)`

Reply:
>  all completely unrelated that we take the mean of the domain with self.prior_linearized(X - self._X_mean)

Yep unrelated to that.  Here's some [explanation on the MMD statistic](https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html), I don't think I linked that already.  

You can use this MMD statistic thing to tell if two sets of samples come from the same distribution.  It's a kernel test, can be applied to samples from any multivariate situation, doesn't have to be applied to samples drawn from other kernel methods (sorry for the inception-ness we have here in this case).  So here, sets of samples from the periodic HSGP prior and samples from the periodic GP prior should look like they came from the same distribution.  The kernel the MMD stat uses (expquad) is unrelated to the kernels from the GPs, it's its own thing.   

Actually I just wrote a long reply and deleted it cuz I found this.  Try running this to see why the MMD statistic started passing after removing the mean:

```python
with pm.Model() as model:
    
    eta = pm.Exponential("eta", lam=1.0)
    ell = pm.InverseGamma("ell", mu=2, sigma=1)
    per = 1.0
    cov_func = pm.gp.cov.Periodic(1, ls=ell, period=per)
    
    hsgp = pm.gp.HSGP(m=[500], cov_func=cov_func)
    f_hsgp = hsgp.prior("f_hsgp", X=x[:, None])
    
    gp = pm.gp.Latent(cov_func=cov_func)
    f_gp = gp.prior("f_gp", X=x[:, None])
    
    idata = pm.sample_prior_predictive(1000)

fig, axs = plt.subplots(1, 2, figsize=(15, 5))
f_hsgp = az.extract(idata.prior, var_names="f_hsgp")
axs[0].plot(f_hsgp.data);
axs[0].set_ylim([-3, 3])

f_gp = az.extract(idata.prior, var_names="f_gp")
axs[1].plot(f_gp.data);
axs[1].set_ylim([-3, 3]);
```

The HSGP prior samples don't have as much level shift up and down, or maybe a lower overall variance.  Maybe something is still off in the implementation, or further down in some special function like `iv`?  I also wonder now if the Stan or Numpyro implementations also show this behavior if you take draws from their priors?  I'll probably leave this over the weekend and come back to it next week, but I can help you run this stuff down.  

I think if Stan/numpyro show the same behavior, we can just mark the test `xfail` with a (fairly long probably) explanation.  I don't think the mean subtraction thing I was doing is legit anymore...  Maybe worth reaching out to the paper authors in that case?

Reply:
Running that code, I got the pytensor warning `UserWarning: Optimization Warning: The Op iv does not provide a C implementation. As well as being potentially slow, this also disables loop fusion.` Could that be related? I don't know if there's a way to run `sample_prior_predictive` using jax. Given the coefficients are`c * pt.iv(J, a) / pt.exp(a)`, maybe all the samples are out by `pt.iv(J, a)`? Wild guess though

I'll write you a snippet for the numpyro now, Stan would take a bit longer for me

Reply:
Here's the numpyro model:

```python
import jax
import jax.numpy as jnp
from tensorflow_probability.substrates import jax as tfp
import numpyro
import numpyro.distributions as dist
from numpyro.infer import Predictive

def modified_bessel_first_kind(v, z):
    v = jnp.asarray(v, dtype=float)
    return jnp.exp(jnp.abs(z)) * tfp.math.bessel_ive(v, z)

def diag_spectral_density_periodic(alpha, length, M):
    a = length ** (-2)
    J = jnp.arange(0, M)
    c = jnp.where(J > 0, 2, 1)
    q2 = (c * alpha**2 / jnp.exp(a)) * modified_bessel_first_kind(J, a)
    return q2

def eigenfunctions_periodic(x, w0, M):
    m1 = jnp.tile(w0 * x[:, None], M)
    m2 = jnp.diag(jnp.arange(M, dtype=jnp.float32))
    mw0x = m1 @ m2
    cosines = jnp.cos(mw0x)
    sines = jnp.sin(mw0x)
    return cosines, sines

def model(X, y=None):
    per = 1.0
    w0 = (jnp.pi * 2 / per)
    M = 500

    # alpha = numpyro.sample("alpha", dist.HalfNormal(1.0)) # this is the variance, setting to 1
    alpha = 1.0
    # numpyro does not have mu, sigma parametrisation, so change the pymc prior to compare
    ell = numpyro.sample("ell", dist.InverseGamma(2, 1))

    q2 = diag_spectral_density_periodic(alpha, ell, M)
    cosines, sines = eigenfunctions_periodic(X, w0, M)

    with numpyro.plate("cos_basis", M):
        beta_cos = numpyro.sample("beta_cos", dist.Normal(0, 1))
    
    with numpyro.plate("sin_basis", M - 1):
        beta_sin = numpyro.sample("beta_sin", dist.Normal(0, 1))

    zero = jnp.array([0.0])
    beta_sin = jnp.concatenate((zero, beta_sin))

    f = numpyro.deterministic("f_hsgp", cosines @ (q2 * beta_cos) + sines @ (q2 * beta_sin))

    # nominal likelihood, ignored for prior prediction
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    with numpyro.plate("obs", X.shape[0]):
        numpyro.sample("y", dist.Normal(f, sigma), obs=y)

prior_predictive = Predictive(model, num_samples=1000)
prior_predictions = prior_predictive(jax.random.PRNGKey(0), X=jnp.squeeze(X1))
```

Visually, this has the same variance as our pymc implementation

```python
idata_numpyro = az.from_numpyro(prior=prior_predictions)

fig, axs = plt.subplots(1, 2, figsize=(15, 5))
f_hsgp_numpyro = az.extract(idata_numpyro.prior, var_names="f_hsgp")
axs[0].plot(f_hsgp_numpyro.data);
axs[0].set_ylim([-3, 3])

# looks the same as our pymc hsgp
f_hsgp = az.extract(idata.prior, var_names="f_hsgp")
axs[1].plot(f_hsgp.data);
axs[1].set_ylim([-3, 3]);
```

Perhaps related, perhaps not, any idea why the numpyro/Stan models would define the period as `period = 365.25`, `w0 = x.std() * (jnp.pi * 2 / period)` (also [here in Stan](https://github.com/avehtari/casestudies/blob/master/Birthdays/gpbf2.stan#L25-L26))? I don't think this affects our problem because we're looking at the overall variance

Reply:
> UserWarning: Optimization Warning: The Op iv does not provide a C implementation. As well as being potentially slow, this a ...

That's a benign warning, you can ignore. It shouldn't affect results (only speed).

Reply:
I think I found it finally!  So, in your numpyro code I changed alpha from 1 to sqrt(2), and now samples match those from `pm.gp.cov.Periodic`.  I think they must have missed a factor of two somewhere in the paper..?  Another way to see it is to flatten the prior samples together into a pile and plot something like:

```python
# gp samples flattened should match standard normal draws below
plt.hist(np.random.randn(20000), 100, alpha=0.5, density=True, color="k");

# f_pymc samples from pm.gp.cov.Periodic
plt.hist(f_pymc.data.flatten(), 100, alpha=0.5, density=True);

# f_hsgp samples from your numpyro code
plt.hist(f_hsgp.data.flatten(), 100, alpha=0.5, density=True);
```


> Aside: any idea why the numpyro/Stan models would define the period using the standard deviation of the domain: period = 365.25, w0 = x.std() * (jnp.pi * 2 / period) (also [here in Stan](https://github.com/avehtari/casestudies/blob/master/Birthdays/gpbf2.stan#L25-L26))? Just checked and this does not affect our problem because we're looking at the overall varianc

No... not sure.  The period you have in your code and your numpyro code is correct though.

Reply:
Interesting, are we sure on this? Like, is it a hack or do we back the maths of root 2? Are we sure it isn't a coincidence? If so, it's an easy fix: 

```diff
f = (
    self.mean_func(X)
+    + np.sqrt(2) * (
            phi_cos @ (psd * self._beta[:m0])  # type: ignore
             + phi_sin[..., 1:] @ (psd[1:] * self._beta[m0:]
      )  # type: ignore
)
```

If we're happy with it, I'll implement it and put a note, and probably shoot something off to the paper author. What do you think?

Another thing:

Is the `alpha` in the paper a variance or a sigma. If we compare it to the [kernel cookbook](https://www.cs.toronto.edu/~duvenaud/cookbook/), I think `alpha` is `sigma**2`.

I think the numpyro code has an error. They have the line `q2 = (c * alpha**2 / jnp.exp(a)) * modified_bessel_first_kind(J, a)`. Instead, I follow the paper more closely, and do not include `alpha` in the `q` terms (just `q2 = c * pt.iv(J, a) / pt.exp(a)`). If we look at equation B.8 of the paper, It actually has `alpha ** 0.5` as a coefficient – i.e sigma. So I think the numpyro code should say `alpha ** 0.5` or just `alpha` if they're looking at a standard deviation prior.

For reference, the corresponding two lines of Stan code are [this](https://github.com/avehtari/casestudies/blob/master/Birthdays/gpbf2.stan#L65) and [this](https://github.com/avehtari/casestudies/blob/master/Birthdays/gpbasisfun_functions.stan#L9) – looks like `exp(log(alpha)+ ...` that their `sigma_f2` is actually a variance too.

I'm not sure where this leaves us, because we specify cov_funcs like `sigma ** 2 * pm.gp.cov.Periodic(1, ls=ell, period=per)`. It would be annoying to have to explain that specifically for the periodic HSGP case, we should do `sigma * pm.gp.cov.Periodic(1, ls=ell, period=per)`. What do you think?

Reply:
> UPDATE: test does not pass after doing this

Dang I didn't check that.  Will take another look again tomorrow and get back to you.  I'm wondering if the test is in "no way" territory, or if it's saying "kind of close".  Was hoping we got lucky with the factor of 2. 

>  It would be annoying to have to explain that specifically for the periodic HSGP case, we should do sigma * pm.gp.cov.Periodic(1, ls=ell, period=per)

For sure, I would hope that HSGP periodic can be a drop in replacement for regular periodic, without having to rethink your priors.